pkg/controller/elasticsearch/elasticsearch_controller.go

```diff
@@ -191,6 +192,11 @@ func (r *ReconcileElasticsearch) Reconcile(ctx context.Context, request reconcil
 		return reconcile.Result{}, tracing.CaptureError(ctx, err)
 	}
 
+	// Requeue if associations are configured but not yet defined
+	if !association.AreConfiguredIfSet(es.GetAssociations(), r.recorder) {
+		return reconcile.Result{}, nil
+	}
```

Comment:
It panics in `monitoringMetricsTargetCaCertSecretVolume` when we try to get the `CASecretName` because the `assocConf` is nil.

```go
func monitoringMetricsTargetCaCertSecretVolume(es esv1.Elasticsearch) volume.SecretVolume {
	assocConf := es.GetMonitoringMetricsAssociation().AssociationConf()
	return volume.NewSecretVolumeWithMountPath(
		assocConf.CASecretName, // <----- HERE
		MonitoringMetricsTargetEsCaCertVolumeName,
		MonitoringMetricsTargetEsCaCertMountPath,
	)
}
```

```
E0616 17:41:04.241541  184548 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)                    
goroutine 801 [running]:                                                                                                                                                                                 
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1a0a280, 0x2a539d0)                                                                                                                                      
        /home/thb/dev/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:74
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)                                                                                                                                          
        /home/thb/dev/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:48
panic(0x1a0a280, 0x2a539d0)                                                                                                                                                                              
        /home/thb/dev/tools/go/src/runtime/panic.go:965
pkg/controller/elasticsearch/stackmon.monitoringMetricsTargetCaCertSecretVolume(...)
        pkg/controller/elasticsearch/stackmon/volume.go:79
pkg/controller/elasticsearch/stackmon.monitoringVolumes(...)      
        pkg/controller/elasticsearch/stackmon/volume.go:37
pkg/controller/elasticsearch/stackmon.WithMonitoring(...)
        pkg/controller/elasticsearch/stackmon/container.go:54
pkg/controller/elasticsearch/nodespec.BuildPodTemplateSpec(...)
        pkg/controller/elasticsearch/nodespec/podspec.go:94
pkg/controller/elasticsearch/nodespec.BuildStatefulSet(...)      
        pkg/controller/elasticsearch/nodespec/statefulset.go:77
pkg/controller/elasticsearch/nodespec.BuildExpectedResources(...)                        
        pkg/controller/elasticsearch/nodespec/resources.go:64
pkg/controller/elasticsearch/driver.(*defaultDriver).reconcileNodeSpecs(...)                                                                                                                                                                                                  
        pkg/controller/elasticsearch/driver/nodes.go:81
pkg/controller/elasticsearch/driver.(*defaultDriver).Reconcile(...)                                                                
        pkg/controller/elasticsearch/driver/driver.go:261
pkg/controller/elasticsearch.(*ReconcileElasticsearch).internalReconcile(...)
        pkg/controller/elasticsearch/elasticsearch_controller.go:292
```

We do the same thing here:

```
pkg/controller/beat/controller.go
167:    if !association.AreConfiguredIfSet(beat.GetAssociations(), r.recorder) {
168-            return results

pkg/controller/agent/controller.go
156:    if !association.AreConfiguredIfSet(agent.GetAssociations(), r.recorder) {
157-            return results

pkg/controller/apmserver/controller.go
210:    if !association.AreConfiguredIfSet(as.GetAssociations(), r.recorder) {
211-            return reconcile.Result{}, nil
```

Reply:
We can avoid to panic by using `GetCASecretName()`. Then without the `association.AreConfiguredIfSet` check, we create the pod with empty secret name and the reconciliation fails:

```
{FieldValueRequired Required value spec.volumes[14].secret.secretName} 
{FieldValueRequired Required value spec.volumes[16].secret.secretName} 
{FieldValueInvalid Invalid value: \"\": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*') spec.containers[1].env[5].valueFrom.secretKeyRef.name} 
{FieldValueRequired Required value spec.containers[1].env[5].valueFrom.secretKeyRef.key} 
{FieldValueNotFound Not found: \"es-monitoring-metrics-target-certs\" spec.containers[1].volumeMounts[2].name} 
```

> Should it not be OK to deploy Elasticsearch even if the monitoring clusters are not ready yet or the association not configured?

With or without restarting the Elasticsearch container when it's ready/configured?
I agree it would be good to have this but it doesn't seem trivial.
I see 3 options:

When monitoring clusters are not ready yet or the association not configured?
1) wait and deploy ES with the sidecars
2) deploy ES and when it's ready recreate the pod with the sidecars
3) don't wait, deploy ES with the sidecars by "precomputing" the association conf

We do (1). (2) involves restarting the Elasticsearch container to reconfigure the pod with the sidecars.

I'm not sure of the complexity of (3). Assoc conf info are `{"authSecretName":"test-c-m2-beat-mon-user","authSecretKey":"a-test-c-m2-beat-mon-user","caCertProvided":true,"caSecretName":"test-es-mon-c-m2-ca","url":"https://m2-es-http.c.svc:9200","version":"7.14.0"}`.

I do not have a strong opinion between (1) and (2).

Reply:
If it is complicated it is maybe not worth it. We can also revisit this later once we have the basic functionality merged.